### PR TITLE
Refactoring of the sc-meta framework version management

### PR DESCRIPTION
**Pull request by BiancaIalangi**
A new FrameworkVersion type is carrying the weight. This makes code cleaner and safer.

### CODE REVIEW

File: framework/meta/src/cmd/standalone/upgrade/upgrade_print.rs [hunk 1]
```diff
@@ -10,7 +13,9 @@ pub fn print_upgrading(dir: &RelevantDirectory) {
         println!(
             "\n{}",
             format!(
-                "Upgrading from {from_version} to {to_version} in {}\n",
+                "Upgrading from {} to {} in {}\n",
+                from_version.version,
```
Comment (andrei-marinica) [inline framework/meta/src/cmd/standalone/upgrade/upgrade_print.rs:17]: We can implement the `Display` trait implemented for FrameworkVersion, and then have from_version/to_version printed directly.

File: framework/derive/src/format/format_version.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+use proc_macro::{quote, Literal};
+
+use crate::format::format_tokenize;
+
+pub fn format_version(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
```
Comment (andrei-marinica) [inline framework/derive/src/format/format_version.rs:5]: I would rename this to `version_triple`, or `version_tuple`,  or `semver_tuple`, whichever you prefer. Same name for the file.

File: framework/derive/src/format/format_version.rs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            _ => panic!("Tokentree does not match with the requirements"),
+        })
+        .collect()
+}
+
+fn u64_literal_from_str(s: &str) -> proc_macro::TokenTree {
+    proc_macro::TokenTree::Literal(Literal::u64_suffixed(
+        s.parse().expect("failed to parse token as u64"),
+    ))
+}
```
Comment (andrei-marinica) [inline framework/derive/src/format/format_version.rs:37]: Call a `cargo fmt` to clean everything up.

File: framework/derive/src/lib.rs [hunk 4]
```diff
@@ -62,3 +62,8 @@ pub fn managed_vec_item_derive(input: proc_macro::TokenStream) -> proc_macro::To
 pub fn format_receiver_args(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
     format::format_receiver_args_macro(input)
 }
+
+#[proc_macro]
```
Comment (andrei-marinica) [inline framework/derive/src/lib.rs:66]: Same rename as above.

File: framework/meta/src/cmd/standalone/template/contract_creator.rs [hunk 5]
```diff
@@ -31,10 +32,10 @@ fn target_from_args(args: &TemplateArgs) -> ContractCreatorTarget {
 
 pub(crate) fn get_repo_version(args_tag: &Option<String>) -> RepoVersion {
     if let Some(tag) = args_tag {
-        assert!(validate_template_tag(tag), "invalid template tag");
+                assert!(validate_template_tag(tag), "invalid template tag");
         RepoVersion::Tag(tag.clone())
     } else {
-        RepoVersion::Tag(LAST_TEMPLATE_VERSION.to_string())
+        RepoVersion::Tag(LAST_TEMPLATE_VERSION.version.to_string())
```
Comment (andrei-marinica) [inline framework/meta/src/cmd/standalone/template/contract_creator.rs:38]: If you implement `Display` for `FrameworkVersion`, you can call to_string() on it directly.

File: framework/meta/src/cmd/standalone/upgrade/upgrade_selector.rs [hunk 6]
```diff
@@ -69,41 +78,32 @@ fn upgrade_function_selector(dir: &RelevantDirectory) {
...
-        Some((_, "0.45.0")) => {
-            upgrade_to_45_0(dir);
-        },
-        Some((from_version, to_version)) => {
-            version_bump_in_cargo_toml(&dir.path, from_version, to_version);
-        },
-        None => {},
+    if let Some((from_version, to_version)) = dir.upgrade_in_progress {
+        match to_version.version.to_string().as_str() {
+            "0.31.0" => upgrade_to_31_0(dir),
```
Comment (andrei-marinica) [inline framework/meta/src/cmd/standalone/upgrade/upgrade_selector.rs:83]: Great place to put our new macro to use, and match on the FrameworkVersion directly.

File: framework/meta/src/cmd/standalone/upgrade/upgrade_selector.rs [hunk 7]
```diff
@@ -69,41 +78,32 @@ fn upgrade_function_selector(dir: &RelevantDirectory) {
...
 fn upgrade_post_processing(dir: &RelevantDirectory, settings: &UpgradeSettings) {
-    match dir.upgrade_in_progress {
-        Some((_, "0.28.0")) | Some((_, "0.29.0")) | Some((_, "0.30.0")) | Some((_, "0.31.0"))
-        | Some((_, "0.32.0")) | Some((_, "0.33.0")) | Some((_, "0.34.0")) | Some((_, "0.35.0"))
-        | Some((_, "0.36.0")) | Some((_, "0.37.0")) | Some((_, "0.40.0")) | Some((_, "0.41.0"))
-        | Some((_, "0.42.0")) | Some((_, "0.43.0")) | Some((_, "0.44.0")) | Some((_, "0.45.2"))
-        | Some((_, "0.46.0")) => {
+    if let Some((_, to_version)) = dir.upgrade_in_progress {
+        if [
+            "0.28.0", "0.29.0", "0.30.0", "0.31.0", "0.32.0", "0.33.0", "0.34.0", "0.35.0",
```
Comment (andrei-marinica) [inline framework/meta/src/cmd/standalone/upgrade/upgrade_selector.rs:95]: We should extract this to a const array, same as the big version list, using the same macro. Can also reside in version_history.rs. You can call it something like `CHECK_AFTER_UPGRADE_TO`

File: framework/meta/src/cmd/standalone/upgrade/upgrade_selector.rs [hunk 8]
```diff
@@ -69,41 +78,32 @@ fn upgrade_function_selector(dir: &RelevantDirectory) {
...
+            "0.36.0", "0.37.0", "0.40.0", "0.41.0", "0.42.0", "0.43.0", "0.44.0", "0.45.2",
+            "0.46.0",
+        ]
+        .contains(&to_version.version.to_string().as_str())
+        {
             print_post_processing(dir);
             cargo_check(dir, settings);
-        },
-        Some((_, "0.39.0")) => {
+        } else if to_version.version.to_string().as_str() == "0.39.0" {
```
Comment (andrei-marinica) [inline framework/meta/src/cmd/standalone/upgrade/upgrade_selector.rs:103]: Macro, and then compare versions.

File: framework/meta/src/folder_structure/relevant_directory.rs [hunk 9]
```diff
@@ -33,7 +33,7 @@ pub enum DirectoryType {
 pub struct RelevantDirectory {
     pub path: PathBuf,
     pub version: VersionReq,
-    pub upgrade_in_progress: Option<(&'static str, &'static str)>,
+    pub upgrade_in_progress: Option<(&'static FrameworkVersion, &'static FrameworkVersion)>,
```
Comment (andrei-marinica) [inline framework/meta/src/folder_structure/relevant_directory.rs:36]: I think it's possible to drop the `'static` at this point (not sure why it was there in the first place).

File: framework/meta/src/folder_structure/relevant_directory.rs [hunk 10]
```diff
@@ -83,26 +83,26 @@ impl RelevantDirectories {
...
     ) -> impl Iterator<Item = &RelevantDirectory> {
         self.0
             .iter()
-            .filter(move |dir| dir.version.semver == version)
+            .filter(move |dir| dir.version.semver == *version)
     }
 
     /// Marks all appropriate directories as ready for upgrade.
-    pub fn start_upgrade(&mut self, from_version: &'static str, to_version: &'static str) {
+    pub fn start_upgrade(&mut self, from_version: &'static FrameworkVersion, to_version: &'static FrameworkVersion) {
```
Comment (andrei-marinica) [inline framework/meta/src/folder_structure/relevant_directory.rs:103]: 'static lifetime also not required here, I think

File: framework/meta/src/version.rs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+impl PartialEq for FrameworkVersion {
+    fn eq(&self, other: &Self) -> bool {
+        self.version == other.version
+    }
+}
+
+pub fn is_sorted(versions: &[FrameworkVersion]) -> bool {
+    versions
+        .windows(2)
+        .all(|window| (window[0].cmp(&window[1])).eq(&Ordering::Less))
```
Comment (andrei-marinica) [inline framework/meta/src/version.rs:60]: Nice! I didn't know this feature.

One thing only: I'm pretty sure you can write window[0] < window[1] directly.

File: framework/meta/src/version_history.rs [hunk 12]
```diff
@@ -1,113 +1,109 @@
...
 /// Should be edited every time a new version of the framework is released.
-pub const LAST_VERSION: &str = "0.46.1";
+pub const LAST_VERSION: FrameworkVersion = framework_version!(0.46.1);
 
 /// Indicates where to stop with the upgrades.
-pub const LAST_UPGRADE_VERSION: &str = LAST_VERSION;
+pub const LAST_UPGRADE_VERSION: FrameworkVersion = LAST_VERSION;
 
-pub const LAST_TEMPLATE_VERSION: &str = LAST_VERSION;
+pub const LAST_TEMPLATE_VERSION: FrameworkVersion = framework_version!(0.45.2);
```
Comment (andrei-marinica) [inline framework/meta/src/version_history.rs:11]: This one should also be 0.46.1, LAST_VERSION

File: framework/meta/src/version_history.rs [hunk 13]
```diff
@@ -1,113 +1,109 @@
...
+    0.45.2,
+    0.46.0,
+    0.46.1,
 ];
 
-/// We started supporting contract templates with version 0.43.0.
-pub fn template_versions() -> &'static [&'static str] {
-    &VERSIONS[33..]
-}
+pub const LOWER_VERSION_WITH_TEMPLATE_TAG: &FrameworkVersion = &VERSIONS[33];
```
Comment (andrei-marinica) [inline framework/meta/src/version_history.rs:62]: We no longer need to reference the index in the big list, we can write it explicitly `framework_version(...)`

File: framework/meta/src/version_history.rs [hunk 14]
```diff
@@ -1,113 +1,109 @@
...
+pub const TEMPLATE_VERSIONS_WITH_AUTOGENERATED_WASM: &FrameworkVersion = &VERSIONS[40];
```
Comment (andrei-marinica) [inline framework/meta/src/version_history.rs:63]: LOWER_VERSION_WITH_AUTOGENERATED_WASM
or
FIRST_VERSION_WITH_AUTOGENERATED_WASM

Comment (andrei-marinica) [reply framework/meta/src/version_history.rs:63]: same for the others

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

File: framework/meta/src/folder_structure/relevant_directory.rs (hunk 10 above)
Comment (BiancaIalangi) [reply framework/meta/src/folder_structure/relevant_directory.rs]: Can't remove 'static lifetime due to error "does not live long enough"

File: framework/meta/src/version_history.rs [hunk 15]
```diff
@@ -1,113 +1,130 @@
...
-pub fn validate_template_tag(tag: &str) -> bool {
-    template_versions().iter().any(|&tt| tt == tag)
-}
+/// We started supporting contract templates with version 0.43.0.
+pub fn validate_template_tag(tag_str: &str) -> bool {
+    let tag: FrameworkVersion = FrameworkVersion::from_string_template(tag_str);
 
-pub fn template_versions_with_autogenerated_wasm() -> &'static [&'static str] {
-    &VERSIONS[40..]
+    tag >= LOWER_VERSION_WITH_TEMPLATE_TAG && tag <= LAST_VERSION
```
Comment (andrei-marinica) [inline framework/meta/src/version_history.rs:91]: Good catch! We should probably validate all framework versions when parsing.
Then we no longer need conditions like `tag <= LAST_VERSION`.
Maybe add a function like `parse_known_version`

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Review by mihaicalinluca — APPROVED** [review]

**Pull request merged**